File: 20.valid-parentheses/solution.cpp
#include <iostream>
#include <stack>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> con;
        for(string::iterator it=s.begin(); it!=s.end(); it++) {
            switch (*it) {
                case '{':
                case '(':
                case '[': con.push(*it); break;
                case ')': if(con.empty()||con.top()!='(') return false; con.pop();break;
                case '}': if(con.empty()||con.top()!='{') return false; con.pop();break;
                case ']': if(con.empty()||con.top()!='[') return false; con.pop();break;
            }
        }
        return con.empty();
    }
};

File: 20.valid-parentheses/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut st: Vec<char> = vec![];
        for c in s.chars() {
            // println!("{}", c);
            match c {
                '(' | '[' | '{' => st.push(c),
                ')' | ']' | '}' => {
                    if let Some(last_el) = st.pop() {
                        if last_el != Solution::get_pair(c) {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
                _ => {}
            };
            // println!("{:?}", st);
        }

        st.len() == 0
    }

    pub fn get_pair(c: char) -> char {
        match c {
            ')' => '(',
            '}' => '{',
            ']' => '[',
            _ => '1'
        }
    }
}
